cypress-io/cypress
# fix: Time out unmatched prerequests in proxy to avoid leaking memory

### User facing changelog
No user facing impact.

### Additional details
Refactored the prerequest proxy logic to be easier to follow. It now uses plain javascript objects with a string key to identify prerequests and requests, and creates fewer closure functions.

It also sweeps unmatched prerequests more than 10s old, and doesn't retain references to requests in the `metrics` object, ensuring that we're not leaking memory. **I have not verified that this occurs in scenarios that https://github.com/cypress-io/cypress/pull/22460 doesn't already address,** but I consider the refactor and minor performance gain to be worthwhile even with that PR making it in.

This should be considered a companion to, but not replacement for https://github.com/cypress-io/cypress/pull/22460. That PR has better performance in the case of data urls by catching them much sooner. This refactor and update is aimed more at readability, and ensuring we don't leak memory in other situations (which may or may not be happening).

### Steps to test
This PR does not require manual testing. Our proxy logic is extensively tested with e2e and unit tests.

Run the unit tests for the `proxy` package. Existing tests are ~1% faster according to mocha, which is hard to notice because they're already quite fast. (this PR does add a couple of additional tests covering existing and new functionality, so total test time is longer than before on my machine).

### How has the user experience changed?


### PR Tasks
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
